File: bot-api/java/src/main/java/dev/robocode/tankroyale/botapi/events/NextTurnCondition.java
package dev.robocode.tankroyale.botapi.events;

import dev.robocode.tankroyale.botapi.IBaseBot;
import dev.robocode.tankroyale.botapi.IBot;

/**
 * Prebuilt condition that can be used for waiting for the next turn.
 * See {@link Condition} and {@link IBot#waitFor} for more information.
 *
 * <p>Here is an example of how to use this condition:
 * <script src="../../../../../prism.js"></script>
 * <pre><code class="language-java">
 *  public class MyBot extends Bot {
 *    public void run() {
 *      while (isRunning()) {
 *        ...
 *        setFire(1.0);
 *        ...
 *        waitFor(new NextTurnCondition(this)); // wait for the next turn before continuing
 *      }
 *    }
 * }
 * </code></pre>
 */
@SuppressWarnings("unused")
public final class NextTurnCondition extends Condition {

    private final IBaseBot baseBot;
    private final int creationTurnNumber;

    /**
     * Constructor for initializing a new instance of the NextTurnCondition class.
     *
     * @param baseBot is your bot instance, typically {@code this} instance, used for determining
     *                the current turn of the battle with the {@link #test} method.
     */
    public NextTurnCondition(IBaseBot baseBot) {
        super("NextTurnCondition");

        this.baseBot = baseBot;
        this.creationTurnNumber = baseBot.getTurnNumber();
    }

    /**
     * This method tests if the turn number has changed since we created this condition.
     *
     * @return {@code true} if the current turn number is greater than the initial turn number,
     * when this condition was created; {@code false} otherwise.
     */
    @Override
    public boolean test() {
        return baseBot.getTurnNumber() > creationTurnNumber;
    }
}

File: bot-api/dotnet/api/src/events/NextTurnCondition.cs
namespace Robocode.TankRoyale.BotApi.Events;

/// <summary>
/// Prebuilt condition that can be used for waiting for the next turn.
/// See <see cref="Condition"/> and <see cref="IBot.WaitFor(Condition)"/> for more information.
/// </summary>
/// <example>
/// <code>
/// public class MyBot : Bot
/// {
///   public void Run()
///   {
///     while (IsRunning)
///     {
///       ...
///       SetFire(1.0);
///       ...
///       WaitFor(new TurnCompleteCondition(this)); // wait for the next turn before continuing
///     }
///   }
/// }
/// </code>
/// </example>
public sealed class NextTurnCondition : Condition
{
    private readonly IBaseBot baseBot;
    private readonly int creationTurnNumber;

    /// <summary>
    /// Constructor for initializing a new instance of the NextTurnCondition class.
    /// </summary>
    /// <param name="baseBot">baseBot is your bot instance, typically <c>this</c> instance, used
    /// for determining the current turn of the battle with the <see cref="Test"/> method</param>
    public NextTurnCondition(IBaseBot baseBot) : base("NextTurnCondition")
    {
        this.baseBot = baseBot;
        creationTurnNumber = baseBot.TurnNumber;
    }

    /// <summary>
    /// This method tests if the turn number has changed since we created this condition.
    /// </summary>
    /// <returns><c>true</c> if the current turn number is greater than the initial turn number,
    /// when this condition was created; <code>false</code> otherwise.</returns>
    public override bool Test()
    {
        return baseBot.TurnNumber > creationTurnNumber;
    }
}